File: C++ Implementations/src/linear_structures/queue.cpp
#include "linear_structures/queue.h"

// ==== Helper Functions ====
Node* Queue::get_node(int index) {
    Node* cursor = head;
    for (int i = 0; i < index; i++) {
        cursor = cursor->get_next();
    }
    return cursor;
}

// ==== Standard Methods ====
Queue::~Queue() {
    // Destructor for deallocating all the DAM used for SLL nodes in the queue.
    Node* cursor = head;
    while (cursor != nullptr) {
        Node* next = cursor->get_next();
        delete cursor;
        cursor = next;
    }
}

std::ostream& operator<<(std::ostream& os, const Queue& print_queue) {
    // Overloaded Extraction operator to display the queue.
    
    // If the queue is empty.
    if (!print_queue.head) {
        os << "None <= Head\nNone <= Tail\n";
    }
    // If the queue is not empty.
    else {
        Node* cursor = print_queue.head;
        while (cursor) {
            if (cursor == print_queue.head) {
                os << cursor->get_value() << " <= Head" << std::endl;
            }
            else if (cursor == print_queue.tail) {
                os << cursor->get_value() << " <= Tail" << std::endl;
            }
            else {
                os << cursor->get_value() << std::endl;
            }
            cursor = cursor->get_next();
        }
    }
    return os;
}

// ==== Member Functions ====
void Queue::enqueue(int value) {
    // Enter a new element to the Queue.

    // DAM for a new global node that comprises the Queue.
    Node* new_node = new Node(value);

    // If the Queue is empty.
    if (head == nullptr) {
        head = new_node;
        tail = new_node;
    }
    // If the Queue is not empty.
    else {
        tail->set_next(new_node);
        tail = new_node;
    }

    // Updating the no of elements in teh queue.
    no_of_elements += 1;
}

int Queue::dequeue() {
    // Removes the first elements in the queue and returns it.

    if (!head) {
        throw std::out_of_range("The queue is empty and has no elements that can be removed.");
    }

    // If the queue has only one element left.
    Node* remove_node = nullptr;
    if (no_of_elements == 1) {
        remove_node = head;
        head = nullptr;
        tail = nullptr;
    }
    // If the queue has more than one element left.
    else {
        remove_node = head;
        head = remove_node->get_next();
    }

    // Updating the link of the remove node
    remove_node->set_next(nullptr);

    // Updating the no of elements
    no_of_elements -= 1;
    return remove_node->get_value();
}

int Queue::peek() const {
    // Returns the top most element in the Queue.
    if (!head) {
        throw std::out_of_range("There are no elements in the list for peek.");
    }
    return head->get_value();
}

bool Queue::is_empty() const {
    // Returns a boolean for if the Queue is empty.
    if (no_of_elements == 0)
        return true;
    return false;
}

void Queue::delete_queue() {
    // Delete the entire queue.
    Node* cursor = head;
    while (cursor) {
        Node* next = cursor->get_next();
        delete cursor;
        cursor = next;
    }

    // Resetting the Pointers and Counter
    no_of_elements = 0;
    head = nullptr;
    tail = nullptr;
}

File: Python Implementations/linear_structures/queue.py
from linear_structures.sll import SLL, Node
from typing import Any


class Queue:
    """Implements the Queue Data Structure using a Linked List."""

    # ==== Standard Methods ====
    def __init__(self) -> None:
        
        self.head: Node | None = None
        self.tail: Node | None = None
        self.no_of_elements: int = 0
        self.queue = SLL()

    def __str__(self) -> str:
        """Provides a string representation for the Queue."""
        
        repr_str: str = ""

        # If the queue is empty.
        if not self.head:
            repr_str = "None <= Head\nNone <= Tail\n"
        # If the queue is not empty.
        else:
            for node in self.queue:
                if node == self.head:
                    repr_str += f"{node} <= Head\n"
                elif node == self.tail:
                    repr_str += f"{node} <= Tail\n"
                else:
                    repr_str += f"{node}\n"
        
        return repr_str
    
    # ==== Methods of the Queue =====
    def enqueue(self, value: Any) -> None:
        """Adds a new element to the end of the queue."""

        # If the queue is currently empty.
        if not self.head:
            self.queue.append(value)
            self.head = self.queue.head
            self.tail = self.queue.tail
        # If the queue is not empty
        else:
            self.queue.append(value)
            self.tail = self.queue.tail

        # Updating the no of elements in the queue
        self.no_of_elements += 1

    def dequeue(self) -> Node:
        """Removes the first node for the queue and returns it."""
        
        remove_node: Node

        # If the queue is empty
        if not self.head:
            raise IndexError("There are no nodes in the queue that can be popped.")
        # If the queue is not empty
        else:
            remove_node = self.queue.remove(0)
            self.head = self.queue.head

        # Updating the no of elements
        self.no_of_elements -= 1
        return remove_node
    
    def peek(self) -> Node:
        """Returns the top most node in the queue."""
        
        # If the queue is empty.
        if not self.head:
            raise IndexError("The queue is currently empty, enter new nodes.")
        # If the queue is not empty.
        else:
            return self.head
        
    def is_empty(self) -> bool:
        """Returns a boolean after checking for an empty queue."""
        
        if not self.head:
            return True
        return False
    
    def delete(self) -> None:
        """Deletes the current queue."""
        
        self.queue.clear()
        self.head = None
        self.tail = None
        self.no_of_elements = 0
